crates/matrix-sdk/src/sliding_sync.rs

```diff
@@ -228,44 +294,93 @@ pub struct UpdateSummary {
+    /// Customize the homeserver for sliding sync only
+    #[builder(setter(strip_option), default)]
     homeserver: Option<Url>,
 
-    #[builder(private)]
+    /// The client this sliding sync will be using
     client: Client,
-
-    // ------ Internal state
     #[builder(private, default)]
-    pos: PosState,
-
-    /// The views of this sliding sync instance
+    views: Vec<SlidingSyncView>,
     #[builder(private, default)]
-    pub views: ViewsList,
-
+    extensions: Option<ExtensionsConfig>,
     #[builder(private, default)]
-    subscriptions: RoomsSubscriptions,
+    subscriptions: BTreeMap<OwnedRoomId, v4::RoomSubscription>,
+}
 
-    #[builder(private, default)]
-    unsubscribe: RoomUnsubscribe,
+impl SlidingSyncConfig {
+    pub async fn build(self) -> Result<SlidingSync> {
+        let SlidingSyncConfig {
+            homeserver,
+            storage_key,
+            client,
+            mut views,
+            extensions,
+            subscriptions,
+        } = self;
+        let rooms;
+        if let Some(storage_key) = storage_key.as_ref() {
+            if let Some(mut f) = client
+                .store()
+                .get_custom_value(storage_key.as_bytes())
+                .await?
+                .map(|v| serde_json::from_slice::<FrozenSlidingSync>(&v))
+                .transpose()?
+            {
+                for view in views.iter_mut() {
+                    if let Some(frozen_view) = f.views.remove(&view.name) {
+                        view.set_from_cold(frozen_view);
+                    }
+                }
 
-    /// The rooms details
-    #[builder(private, default)]
-    rooms: RoomsMap,
+                rooms = f.rooms.into_iter().map(|(k, v)| (k, v.into())).collect();
+            } else {
+                rooms = Default::default();
+            }
+        } else {
+            rooms = Default::default();
+        };
 
-    #[builder(private, default)]
-    extensions: Mutable<Option<ExtensionsConfig>>,
+        let rooms = Arc::new(MutableBTreeMap::with_values(rooms));
+        // map the roomsmap into the views:
+        views.iter_mut().for_each(|v| v.rooms = rooms.clone());
+
+        let views = Arc::new(MutableVec::new_with_values(views));
+
+        Ok(SlidingSync {
+            homeserver,
+            client,
+            storage_key,
+
+            views,
+            rooms,
+            extensions: Mutable::new(extensions),
+            failure_count: Default::default(),
+
+            pos: Mutable::new(None),
+            subscriptions: Arc::new(MutableBTreeMap::with_values(subscriptions)),
+            unsubscribe: Default::default(),
+        })
+    }
 }
 
 impl SlidingSyncBuilder {
     /// Convenience function to add a full-sync view to the builder
     pub fn add_fullsync_view(mut self) -> Self {
-        let views = self.views.clone().unwrap_or_default();
-        views.lock_mut().push_cloned(
+        let mut views = self.views.clone().unwrap_or_default();
```

Comment:
I think this can be a `&mut` by using `self.views.get_or_insert_with(Default::default)`, then the clone and assignment below aren't needed.